include/rcutils/types/string_map.h

```diff
@@ -0,0 +1,401 @@
+ */
+RCUTILS_PUBLIC
+RCUTILS_WARN_UNUSED
+rcutils_ret_t
+rcutils_string_map_clear(rcutils_string_map_t * string_map);
+
+/// Set a key value pair in the map, increasing capacity if necessary.
+/**
+ * The capacity will be increased if needed using rcutils_string_map_reserve().
+ * Otherwise it is the same as rcutils_string_map_set_no_resize().
+ *
+ * \see rcutils_string_map_set_no_resize()
+ *
+ * \param[inout] string_map rcutils_string_map_t to be updated
+ * \param[in] key map key, must be null terminated c string
+ * \param[in] value value for given map key, must be null terminated c string
+ * \return `RCUTILS_RET_OK` if successful, or
+ * \return `RCUTILS_RET_INVALID_ARGUMENT` for invalid arguments, or
+ * \return `RCUTILS_RET_BAD_ALLOC` if memory allocation fails, or
+ * \return `RCUTILS_RET_STRING_MAP_INVALID` if the string map is invalid, or
+ * \return `RCUTILS_RET_ERROR` if an unknown error occurs
+ */
+RCUTILS_PUBLIC
+RCUTILS_WARN_UNUSED
+rcutils_ret_t
+rcutils_string_map_set(rcutils_string_map_t * string_map, const char * key, const char * value);
+
+/// Set a key value pair in the map but only if the map has enough capacity.
+/**
+ * If the map already contains the given key, the existing value will be
+ * replaced with the given value.
+ * If the map does not contain the given key, and the map has additional
+ * unused capacity, then it will store the given key and value in the map.
+ * If there is no unused capacity in the map, then RCUTILS_RET_NOT_ENOUGH_SPACE
+ * is returned.
+ *
+ * The given key and value c strings are copied into the map, and so storage is
+ * allocated for them in the map when this function is called if necessary.
+ * The storage allocated for this purpose is reclaimed either when
+ * rcutils_string_map_fini() is called on this map or when using this function
+ * or rcutils_string_map_unset().
+ *
+ * Any allocation that occurs in this functions uses the allocator of the map,
+ * which is given when the map is initialized in rcutils_string_map_init().
+ *
+ * \param[inout] string_map rcutils_string_map_t to be updated
+ * \param[in] key map key, must be null terminated c string
+ * \param[in] value value for given map key, must be null terminated c string
+ * \return `RCUTILS_RET_OK` if successful, or
+ * \return `RCUTILS_RET_INVALID_ARGUMENT` for invalid arguments, or
+ * \return `RCUTILS_RET_BAD_ALLOC` if memory allocation fails, or
+ * \return `RCUTILS_RET_STRING_MAP_INVALID` if the string map is invalid, or
+ * \return `RCUTILS_RET_NOT_ENOUGH_SPACE` if map is full, or
+ * \return `RCUTILS_RET_ERROR` if an unknown error occurs
+ */
+RCUTILS_PUBLIC
+RCUTILS_WARN_UNUSED
+rcutils_ret_t
+rcutils_string_map_set_no_resize(
+  rcutils_string_map_t * string_map,
+  const char * key,
+  const char * value);
+
+/// Unset a key value pair in the map.
+/**
+ * The key needs to be a null terminated c string.
+ * If the given key is not found, RCUTILS_RET_STRING_KEY_NOT_FOUND is returned.
```

Comment:
I think this is covered by the set function which says that the key and value are copied and managed by the string map:

https://github.com/ros2/rcutils/pull/20/files#diff-d7eafeb8de6275c09bdb50b98d178534R236